File: data-structure/linked/singly_linked.rs
// 定义节点结构体
#[derive(Debug)]
struct Node {
    data: i32,
    next: Option<Box<Node>>,
}

// 定义链表结构体
#[derive(Debug)]
struct LinkedList {
    head: Option<Box<Node>>,
}

impl LinkedList {
    // 初始化一个空链表
    fn new() -> Self {
        LinkedList { head: None }
    }

    // 打印链表
    fn print_list(&self) {
        let mut current = &self.head;
        while let Some(node) = current {
            print!("{} -> ", node.data);
            current = &node.next;
        }
        println!("NULL");
    }

    // 插入新节点到链表尾部
    fn append_node(&mut self, data: i32) {
        let new_node = Box::new(Node {
            data,
            next: None,
        });

        if self.head.is_none() {
            self.head = Some(new_node);
            return;
        }

        let mut current = &mut self.head;
        while let Some(node) = current {
            if node.next.is_none() {
                node.next = Some(new_node);
                break;
            }
            current = &mut node.next;
        }
    }
}

fn main() {
    let mut list = LinkedList::new();

    // 插入节点
    list.append_node(1);
    list.append_node(2);
    list.append_node(3);

    list.print_list(); // 输出：1 -> 2 -> 3 -> NULL
}

/*
jarry@MacBook-Pro linked % rustc singly_linked.rs
jarry@MacBook-Pro linked % ./singly_linked 
1 -> 2 -> 3 -> NULL
*/

File: 01-data-structures/linked/singly_linked.c
#include <stdio.h>
#include <stdlib.h>

// 定义节点结构体
struct Node
{
  int data;
  struct Node *next;
};

// 打印链表
void printList(struct Node *head)
{
  struct Node *temp = head;
  while (temp != NULL)
  {
    printf("%d -> ", temp->data);
    temp = temp->next;
  }
  printf("NULL\n");
}

// 插入新节点到链表尾部
void appendNode(struct Node **head, int data)
{
  struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
  struct Node *last = *head;

  newNode->data = data;
  newNode->next = NULL;

  if (*head == NULL)
  {
    *head = newNode;
    return;
  }

  while (last->next != NULL)
  {
    last = last->next;
  }
  last->next = newNode;
}

int main()
{
  struct Node *head = NULL;

  // 插入节点
  appendNode(&head, 1);
  appendNode(&head, 2);
  appendNode(&head, 3);

  printList(head); // 输出：1 -> 2 -> 3 -> NULL
  return 0;
}

/*
jarry@MacBook-Pro linked % gcc singly_linked.c
jarry@MacBook-Pro linked % ./a.out
1 -> 2 -> 3 -> NULL
 */